CIP-XXX/CIP-XXX.md

```diff
@@ -0,0 +1,1408 @@
+definition for the bit-flip variant, we view a `BuiltinByteString` of length $n$
+as a binary natural number with exactly $8n$ digits, and the value at index $i$
+corresponds to the digits whose place value is either $2^i$ (for the bit-flip
+variant), or $2^{8n - i - 1}$ (for our chosen method). Put another way, under
+the specification for the bit-flip variant, the least significant binary digit
+is first, whereas in our chosen specification, the least significant binary
+digit is last. 
+
+When viewed this way, we can immediately see a potential problem, as by indexing
+a `BuiltinByteString`, we get back a `BuiltinInteger`, which has a numerical
+value as a natural number in the range $[0, 255]$. It would thus be sensible
+that, given a `BuiltinByteString` that is non-empty, if we were to get the
+values at bit indexes $0$ through $7$, and treat them as their corresponding
+place value in a summation, we should obtain the same answer as indexing
+whichever byte those bits come from. We call this notion _conversion agreement_,
+due to its relation to [CIP-121][conversion-cip], which allows us to
+(essentially) view a `BuiltinByteString` as a natural number in base-256.
+Indeed, the case is analogous, with the only difference being that we are
+concerned with the positioning of base-256 digits, not binary ones.
+
+Consider the `BuiltinByteString` whose only byte is $42$, whose representation 
+is as follows:
+
+```
+| Byte index | 0        |
+|------------|----------|
+| Byte       | 00101010 |
+```
+
+Under the bit-flip variant, our bit indexes would be as follows:
+
+```
+| Byte index | 0                             |
+|------------|-------------------------------|
+| Byte       | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 |
+|------------|-------------------------------|
+| Bit index  | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
+```
+
+However, this immediately causes a problem: under this indexing scheme, we imply
+that the $2^2 = 4$ place value is $1$ in $42$'s binary representation, but
+this is not the case. This fails conversion agreement. However, our choice 
+produces the correct answer:
+
+```
+| Byte index | 0                             |
+|------------|-------------------------------|
+| Byte       | 0 | 0 | 1 | 0 | 1 | 0 | 1 | 0 |
+|------------|-------------------------------|
+| Bit index  | 7 | 6 | 5 | 4 | 3 | 2 | 1 | 0 |
+```
+
+Here, the $4$ place value is correctly $0$. Indeed, our choice simply 'extends'
+the bit ordering (in the place value sense) employed by all machine
+architectures for bytes in their sense as natural numbers within a fixed range.
+This was also the choice made by [CIP-58][cip-58], for similar reasons.
+
+We also note that conversion agreement matters for Case 2, in the wider context
+of the interaction between [CIP-121 primitives][conversion-cip] and logical
+XOR. The Argon2 family of hashes use certain inputs (which happen to be numbers)
+both as numbers and also as blocks of binary, which means that any inconsistency
```

Comment:
I think this is a good example, and it leaves me uncertain how this would help us. Is the idea that:
- We somehow make 0-cost conversions from bytestrings to integers to facilitate this case
- Implementors of such algorithms implement their own integer operations on bytestrings based on the given representation
?

Reply:
This means that we don't have to worry about this disagreement in the first place. People can convert integers to bytestrings, fiddle with their bits, then convert back, and expect 'the right results' at the other end.

Reply:
I see, so it's something like my second bullet point: you can convert to a bytestring, do some fiddling that you _expect_ to have consequences for the encoded integer, convert back, and be correct.

Reply:
Yes, exactly. I feel this is an essential feature.

Reply:
I feel like I got slightly lost in the exposition, but are we definitely sure that nobody will want conversion agreement with a _different_ arrangement? For example, with the other endianness.

Reply:
It's not possible to have all of the following:

* Consecutive bit indexes stay consecutive
* Conversion agreement with little endian ordering
* My 1-byte example to make sense

Given that, I had to sacrifice something, and I feel sacrificing conversion agreement with little-endian ordering was the right choice. Do you feel I should state this more explicitly?

Reply:
Yes, that would be helpful. Naively I might have thought that conversion agreement with little-endian ordering (if that's important for some reason!) might be _more_ important than some of the other things. Also, I just found it a bit confusing that you referred to conversion-agreement _unqualified_, when it's conversion agreement with a particular encoding scheme that's chosen as important for... some reason.

Reply:
Yeah, upon re-reading, I agree, it could be more clear. Will add.